File: Sources/Xvideo/Domain/Services/SourceParser.swift
import Foundation

enum SourceParser {
    static func parsePlaybackSources(from item: VodItem) -> [PlaybackSource] {
        let sourceNames = splitSourceNames(item.vodPlayFrom)
        let groups = splitGroups(item.vodPlayURL)

        return groups.enumerated().compactMap { index, group in
            let episodes = parseEpisodes(group)
            guard !episodes.isEmpty else { return nil }
            let name = sourceNames.indices.contains(index) ? sourceNames[index] : "播放源 \(index + 1)"
            return PlaybackSource(id: "\(index)-\(name)", name: name, episodes: episodes)
        }
    }

    static func parseDownloads(from item: VodItem) -> [Episode] {
        parseEpisodes(item.vodDownURL ?? "")
    }

    private static func splitGroups(_ raw: String?) -> [String] {
        guard let raw else { return [] }
        return raw.components(separatedBy: "$$$").filter { !$0.isEmpty }
    }

    private static func splitSourceNames(_ raw: String?) -> [String] {
        splitGroups(raw)
            .flatMap { $0.components(separatedBy: ",") }
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .map { readableSourceName($0) }
    }

    private static func parseEpisodes(_ raw: String) -> [Episode] {
        raw.components(separatedBy: "#").compactMap { pair in
            let pieces = pair.components(separatedBy: "$")
            guard pieces.count >= 2 else { return nil }
            let title = pieces[0].trimmingCharacters(in: .whitespacesAndNewlines)
            let urlString = pieces.dropFirst().joined(separator: "$")
            guard let encoded = urlString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
                  let url = URL(string: encoded) else {
                return nil
            }
            return Episode(title: title.isEmpty ? "未命名" : title, url: url)
        }
    }

    private static func readableSourceName(_ raw: String) -> String {
        switch raw.lowercased() {
        case "lzm3u8":
            return "Xvideo M3U8"
        case "liangzi":
            return "Xvideo 在线"
        case "dyttm3u8":
            return "电影天堂 M3U8"
        case "dytt":
            return "电影天堂 在线"
        default:
            return raw
        }
    }
}

File: src/services/sourceParser.ts
import type { Episode, PlaybackSource, VodItem } from "../types";
import { nilIfBlank } from "./format";

export function parsePlaybackSources(item?: VodItem): PlaybackSource[] {
  if (!item) {
    return [];
  }

  const sourceNames = splitSourceNames(item.vod_play_from);
  const groups = splitGroups(item.vod_play_url);

  return groups.flatMap((group, index) => {
    const episodes = parseEpisodes(group);
    if (episodes.length === 0) {
      return [];
    }
    const name = sourceNames[index] ?? `播放源 ${index + 1}`;
    return [{ id: `${index}-${name}`, name, episodes }];
  });
}

export function parseDownloads(item?: VodItem): Episode[] {
  return parseEpisodes(item?.vod_down_url ?? "");
}

export function episodeCount(item?: VodItem): number {
  return Math.max(0, ...parsePlaybackSources(item).map((source) => source.episodes.length));
}

function splitGroups(raw?: string): string[] {
  return (raw ?? "").split("$$$").filter(Boolean);
}

function splitSourceNames(raw?: string): string[] {
  return splitGroups(raw)
    .flatMap((group) => group.split(","))
    .map((value) => value.trim())
    .filter(Boolean);
}

function parseEpisodes(raw: string): Episode[] {
  return raw.split("#").flatMap((pair) => {
    const pieces = pair.split("$");
    if (pieces.length < 2) {
      return [];
    }

    const title = nilIfBlank(pieces[0]) ?? "未命名";
    const url = pieces.slice(1).join("$").trim();
    if (!url) {
      return [];
    }

    try {
      const normalizedURL = new URL(encodeURI(url)).toString();
      return [{ id: normalizedURL, title, url: normalizedURL }];
    } catch {
      return [];
    }
  });
}
